src/oop/creating_objects_from_classes/hard/q4/MainTest.java

```diff
@@ -0,0 +1,93 @@
+        assertTrue(testClass.hasModifier("main", mainArgs, "public"), incorrectVisibility);
+        assertTrue(testClass.hasModifier("main", mainArgs, "static"), missingStatic);
+        assertTrue(testClass.hasReturnType("main", mainArgs, Void.TYPE), wrongReturnType);
+
+        String incorrectOutput = "Your main method in the %s class does not print the correct output. Make sure that you are printing the balance.";
+        Clause[] clauses = new Clause[]{
+                new StringLiteral("Balance: "),
+                new DoubleLiteral("num1"),
+                new NewLine(),
+                new StringLiteral("Balance: "),
+                new DoubleLiteral("num2"),
+                new Optional(new NewLine())
+        };
+        Object[][] mainParameters = {{new String[0], String[].class}};
+
+        ArrayList<Double> firstRandomValues = new ArrayList<>();
+        ArrayList<Double> secondRandomValues = new ArrayList<>();
+
+        testClass.callMethod("main", mainParameters, clauses, String.format(incorrectOutput, testSavingsName));
+
+        /* Check that the constructor was called and with the correct values */
+        String noConstructorCall = "Your main method did not initialize a %s object.";
+        String incorrectValue = "Your main method did not initialize a %s with the correct %s value.";
+        assertTrue(Savings.isConstructorCalled(), String.format(noConstructorCall, savingsName));
+        assertEquals(initial, Savings.getConstructorInitial(), String.format(incorrectValue, savingsName, field1));
+
+        /* Check that methods were called */
+        String methodNotCalled = "Your main method did not call the %s method%s.";
+        assertTrue(Savings.isGetTotalCalled(), String.format(methodNotCalled, method1Name, ""));
+        assertTrue(Savings.isGetTotalCalledTwice(), String.format(methodNotCalled, method1Name, " a second time"));
+        assertTrue(Savings.isRecieveBonusCalled(), String.format(methodNotCalled, method2Name, ""));
+
+        /* Call method and check randomness */
+        for (int x = 0; x < 1500; x++) {
+            MethodTest m = new MethodTest(BankAccount.class, "main", mainParameters, clauses);
+            m.callMethod();
+            double first = Double.parseDouble(m.getMethodItemByName("num1"));
+            firstRandomValues.add(first - initial);
+            secondRandomValues.add(Double.parseDouble(m.getMethodItemByName("num2")) - first);
+        }
+
+        RandomDouble randomDouble = new RandomDouble(bonusMin, bonusMax);
+        assertTrue(randomDouble.validateRandom(firstRandomValues), "Your first balance is not being correctly generated. Please ensure you are setting the correct initial value.");
+        assertTrue(randomDouble.validateRandom(secondRandomValues), "Your second balance is not being correctly generated. Please ensure you have called the " + method2Name + " method.");
+    }
```

Comment:
I don't think that you need to check the randomness here. The student is not able to tweak the code for the class, so I think you can just record the value that our code generates, and make sure that is the value the test prints. This is more work than we need.